File: contigiousArray.py
# // Time Complexity :O(n) for iterating over the hashmap
# // Space Complexity : O(n) for creating a hashmap of size n;
# // Did this code successfully run on Leetcode : Yes
# // Any problem you faced while coding this : Was iterating using nums instead of range(length(nums)).


# // Your code here along with comments explaining your approach
# We need to find earliest occurance of unique values in the rSum
# after every new running total is added we find if there was already a total previously
# if it is uniue we add it, otherwise we compare current max with length to current index
class Solution:
    def findMaxLength(self, nums: list[int]) -> int:
              
        hMap = {}                           # first occurance map    
        rSum = 0                            # running sum map
        length = 0                          # current max lenght
        hMap[0] = -1                        # edge case for binary

        for i in range(len(nums)):
            if nums[i]==0:                  # 0 -> -1 and 1 -> +1
                rSum -= 1
            else:
                rSum += 1

            if rSum in hMap:                        # if rSum exists in hashmap
                length = max(length, i-hMap[rSum])  # compare current maxLength with hMap[rSum] from current index
            else:
                hMap[rSum] = i                      # add if doesnt exist
        return length
    
arr = [ 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1]
x = Solution().findMaxLength(arr)
print(x)

# Input: nums =  [ 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1]
# Rsum        = 0[-1, 0,-1, 0, 1, 2,-1, 0, 1, 2, 3, 4]

File: Contigious-Array.java
// Time Complexity :O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach.

// Approach: In this we are maintaining a rsum and a count variable. All the unique rsums we are inserting in an hashmap with value as their first instance index.
// Next time when the same rsum is encountered then we will look up in hashmap and then subtract both and update the max.
class Solution {
    public int findMaxLength(int[] nums) {
        //base condition
        if(nums==null || nums.length==0){
            return 0;
        }
        int max=0;
        int rSum=0;
        HashMap<Integer,Integer> map = new HashMap<>();
        //We are inserting the 0 with index as -1 else when in future rsum is zero at that time when it will look up in hashmap it will not find it and we will get incorrect results.
        map.put(0,-1);
        for(int i=0;i<nums.length;i++){
            // If the nums[i] is 0 then we are subtracting from rsum else if it is 1 then we are adding 1 in rsum
            if(nums[i]==0){
                rSum--;
            }
            else{
                rSum++;
            }
            if(!map.containsKey(rSum)){
                map.put(rSum,i);
            }
            max=Math.max(max,i-map.get(rSum));
        }
        //At the end we are returning max as it will be the maximum llength of an subarray
        return max;
    }
}